pkg/testing/runner/provisioner.go

```diff
@@ -0,0 +1,105 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package runner
+
+import (
+	"context"
+
+	"github.com/elastic/elastic-agent/pkg/testing/define"
+)
+
+// Instance represents a provisioned instance.
+type Instance struct {
+	// ID is the identifier of the instance.
+	//
+	// This must be the same ID of the OSBatch.
+	ID string `yaml:"id"`
+	// Name is the nice-name of the instance.
+	Name string `yaml:"name"`
+	// IP is the IP address of the instance.
+	IP string `yaml:"ip"`
+	// Username is the username used to SSH to the instance.
+	Username string `yaml:"username"`
+	// RemotePath is the based path used for performing work on the instance.
+	RemotePath string `yaml:"remote_path"`
+	// Internal holds internal information used by the provisioner.
+	// Best to not touch the contents of this, and leave it be for
+	// the provisioner.
+	Internal map[string]interface{} `yaml:"internal"`
+}
+
+// InstanceProvisioner performs the provisioning of instances.
+type InstanceProvisioner interface {
+	// SetLogger sets the logger for it to use.
+	SetLogger(l Logger)
+
+	// Supported returns true of false if the provisioner supports the given batch.
+	Supported(batch define.OS) bool
+
+	// Provision brings up the machines.
+	//
+	// The provision should re-use already prepared instances when possible.
+	Provision(ctx context.Context, batches []OSBatch) ([]Instance, error)
+
+	// Clean cleans up all provisioned resources.
+	Clean(ctx context.Context, instances []Instance) error
+}
+
+// InstanceProvisionerCreator creates a new provisioner.
+type InstanceProvisionerCreator func(l Logger, cfg Config) InstanceProvisioner
```

Comment:
Absolutely! That is the whole reason for the refactor. I am already working on a LXD provisioner that will allow me to spin up full ubuntu machines locally on an ubuntu host.

Reply:
So we are gonna have tests running both ogc and lxd provisioners at the same time, in the same run, switching between the 2 ? Is that why we need a Factory pattern as defined here ?

Reply:
The original idea was that it would be defined in a registry, but I removed the registry and changed it to just be defined by the creator of the runner. So both `InstanceProvisionerCreator` and `StackProvisionerCreator` can be removed. Follow up commit will remove them.

The idea is that you can use different provisioners, not at the same time in the same run but when running the tests.

There is already work done to switch from OGC to terraform that will be done first with a flag `STACK_PROVISIONER="terraform" mage integration:test`. I am also working on the LXD provider that will allow the same `STACK_PROVISIONER="lxd" mage integration:test`.